apollo-router/src/plugins/authentication/subgraph_auth.rs

```diff
@@ -0,0 +1,455 @@
+            self.access_key_id.clone(),
+            self.secret_access_key.clone(),
+            None,
+            None,
+            "apollo-router",
+        )))
+    }
+}
+
+/// Configuration of the DefaultChainProvider
+#[derive(Clone, JsonSchema, Deserialize, Debug)]
+struct DefaultChainConfig {
+    /// The aws region this chain applies to.
+    region: String,
+    /// The profile name used by this provider
+    profile_name: Option<String>,
+    /// The service you're trying to access, eg: "s3", "vpc-lattice-svcs", etc.
+    service_name: String,
+    /// Specify a role ARN for your credentials.
+    assume_role: Option<AssumeRoleProvider>,
+}
+
+/// Specify a role ARN for your credentials.
+#[derive(Clone, JsonSchema, Deserialize, Debug)]
+struct AssumeRoleProvider {
+    /// Amazon Resource Name (ARN)
+    /// for the role assumed when making requests
+    role_arn: String,
+    /// Uniquely identify a session when the same role is assumed by different principals or for different reasons.
+    session_name: String,
+    /// Unique identifier that might be required when you assume a role in another account.
+    external_id: Option<String>,
+}
+
+/// Configure AWS sigv4 auth.
+#[derive(Clone, JsonSchema, Deserialize, Debug)]
+#[serde(rename_all = "snake_case")]
+enum AWSSigV4Config {
+    Hardcoded(AWSSigV4HardcodedConfig),
+    DefaultChain(DefaultChainConfig),
+}
+
+#[derive(Clone, JsonSchema, Deserialize)]
+#[serde(deny_unknown_fields)]
+enum AuthConfig {
+    #[serde(rename = "aws_sig_v4")]
+    AWSSigV4(AWSSigV4Config),
+}
+
+/// Configure subgraph authentication
+#[derive(Clone, JsonSchema, Deserialize)]
+#[serde(rename_all = "snake_case", deny_unknown_fields)]
+struct Config {
+    /// Configuration that will apply to all subgraphs.
+    #[serde(default)]
+    all: Option<AuthConfig>,
+    #[serde(default)]
+    /// Create a configuration that will apply only to a specific subgraph.
+    subgraphs: HashMap<String, AuthConfig>,
+}
+
+struct SubgraphAuth {
+    signing_params: SigningParams,
+}
+
+#[derive(Clone, Default)]
+struct SigningParams {
+    all: Option<SigningParamsConfig>,
+    subgraphs: HashMap<String, SigningParamsConfig>,
+}
+
+#[derive(Clone)]
+struct SigningParamsConfig {
+    credentials_provider: Option<Arc<dyn ProvideCredentials>>,
+    region: Region,
+    service_name: String,
+}
+
+#[async_trait::async_trait]
+impl Plugin for SubgraphAuth {
+    type Config = Config;
+    async fn new(init: PluginInit<Self::Config>) -> Result<Self, BoxError> {
+        let all = if let Some(config) = &init.config.all {
+            Some(make_signing_params(config).await)
+        } else {
+            None
+        };
+
+        let mut subgraphs: HashMap<String, SigningParamsConfig> = Default::default();
+        for (subgraph_name, config) in &init.config.subgraphs {
+            subgraphs.insert(subgraph_name.clone(), make_signing_params(config).await);
+        }
+
```

Comment:
Would it make sense to attempt to sign a fake or mock request here to test if the configuration is valid. Otherwise, the router starts, but will fail every request to SigV4 subgraphs. I think having it fail during startup makes more sense if it can be caught then. Some situations might still get caught by the runtime error, like if the upstream session provider fails and provides to provide a session.

Reply:
Should the router fail to start / update if the credentials cannot be provided?
I have a fallible `credentials_provider.provide_credentials().await` function i could call at startup, if we wish to do so.

OTOH most of the code "fails open", ie it just doesnt sign the request if it s not able to, so i wonder if that's the direction we want.

@BrynCooke @BlenderDude @bnjjj wdyt?

Reply:
I would prefer to fail when starting if the configuration is wrong but I don't have enough context TBH. But my preference is always be clear with the user that something went wrong instead of trying to do something they don't expect

Reply:
As discussed in huddle:

Action: fail /closed/
- if we don't have credentials to work with, warn and /fail/ plugin startup
- warn if we havent been able to sign a request and fail at subgraph request (controlflow break)
- the subgraph will return a 403 if the signed request is invalid
Assume role: this check is done at runtime, (ie revocation scenarios), we don't fail to startup in this case.

Reply:
Fixed in [3b1e524](https://github.com/apollographql/router/pull/3365/commits/3b1e5247f96f7a028f5503f385d09187945f5fe6)